libraries/lib-files/FileNames.cpp

```diff
@@ -224,13 +232,48 @@ wxString FileNames::LowerCaseAppNameInPath( const wxString & dirIn){
    return dir;
 }
 
-FilePath FileNames::DataDir()
+enum class DirTarget
+{
+    Cache,
+    Config,
+    Data,
+    State
+};
+
+FilePath &gTargetDir(DirTarget target)
+{
+   switch (target)
+   {
+      case DirTarget::Cache: return gCacheDir;
+      case DirTarget::Config: return gConfigDir;
+      case DirTarget::Data: return gDataDir;
+      case DirTarget::State: return gStateDir;
+   }
+   wxASSERT(!"Not all DirTarget cases were implemented.");
+   abort();
+}
+
+const gchar *xdgTargetDir(DirTarget target)
+{
+   switch (target)
+   {
+      case DirTarget::Cache: return g_get_user_cache_dir();
+      case DirTarget::Config: return g_get_user_config_dir();
+      case DirTarget::Data: return g_get_user_data_dir();
+      case DirTarget::State: return g_get_user_state_dir();
+   }
+   wxASSERT(!"Not all DirTarget cases were implemented.");
+   abort();
+}
+
+FilePath userTargetDir(DirTarget target)
 {
+   auto dir = gTargetDir(target);
```

Comment:
You return non const reference from gTargetDir, but here C++ deduce the type as `FilePath`. Is is desired?